### PR TITLE
Add Tasks to quest

### PR DESCRIPTION
**Pull request by lancatlin**
New features
* every stage is a task, once it's requirements fits, it display; once it completed, mark as completed
* change stage.next to stage.requirements, allow multiple stage at the same time

Improvements:
* cli client add debug option to choose whether showing the socket event
* kill the shell process when the socket disconnects
* separate quest image to a base image, easier to create new quest
* add 2 second waiting time a blank line before showing the response
* use config.expiry for jwt expiry time

Bug fixes:
* container not found crash the server -> return error through socket
* saving session at the same time -> only save once by commandController
*

### CODE REVIEW

File: server/src/models/quest.js
Comment (Annie04082020) [inline server/src/models/quest.js:1]: Why use String and [String]? What's the difference?

File: server/src/containers/docker.js [hunk 1]
```diff
@@ -35,16 +35,11 @@ export async function createContainer(name) {
...
   console.log(`Getting container: ${id}`)
-  let container
-  try {
-    container = engine.getContainer(id)
-    if (!(await container.inspect()).State.Running) {
-      await container.start()
-    }
-    return container
-  } catch (error) {
-    throw new Error(`Container not found: ${id}`)
```
Comment (Annie04082020) [inline server/src/containers/docker.js:46]: Why deleted the catch error part?

File: quests/helloworld/Dockerfile
Comment (Annie04082020) [inline quests/helloworld/Dockerfile:1]: Do ./quests/helloworld/home ${HOME} got all this quest need to copy from?

File: packages/client/index.js [hunk 2]
```diff
@@ -1,8 +1,11 @@
 /* eslint-disable no-underscore-dangle */
+import axios from 'axios'
 import { stdin, stdout, exit } from 'process'
 import { io } from 'socket.io-client'
 import { program } from 'commander'
 
+let api = null
```
Comment (Annie04082020) [inline packages/client/index.js:7]: Why let api null?

File: packages/client/index.js [hunk 3]
```diff
@@ -107,6 +122,7 @@ async function main() {
 program
   .option('-s, --session <string>', 'Session ID')
   .option('-c, --create', 'Create a new session')
+  .option('-d, --debug', 'Debug mode')
```
Comment (Annie04082020) [inline packages/client/index.js:125]: What can this debug mode do?

File: server/src/api/controllers/commandController.js [hunk 4]
```diff
@@ -7,7 +7,6 @@ import CommandHandler from '../../game/commandHandler.js'
 export async function newCommand(req, res) {
   console.log('new command:', req.body)
   const { token, command, pwd, output, error, ...additionalData } = req.body
-  console.log('additionalData:', additionalData)
```
Comment (Annie04082020) [inline server/src/api/controllers/commandController.js:10]: So can we still see additional data from log after this? Or it is just move to another place?

File: server/src/api/controllers/sessionController.js [hunk 5]
```diff
@@ -45,20 +46,22 @@ export async function createSession(req, res) {
...
       user: req.user,
       quest,
-      progress,
       containerId: container.id,
     })
 
-    const session = await newSession.save()
-    res.status(201).json(session)
+    const session = new SessionHandler(newSession)
+    session.addNewTasks()
```
Comment (Annie04082020) [inline server/src/api/controllers/sessionController.js:59]: The progress is added in addNewTasks, right

File: quests/helloworld/Dockerfile
Comment (lancatlin) [reply quests/helloworld/Dockerfile:1]: It's enough for this quest, but in other quest, they can copy more file by their needa

File: server/src/api/controllers/commandController.js (hunk 4 above)
Comment (lancatlin) [reply server/src/api/controllers/commandController.js:10]: Oh, I just thought we don't need to see that again. You can see the whole body from req.body after all

File: server/src/api/controllers/sessionController.js (hunk 5 above)
Comment (lancatlin) [reply server/src/api/controllers/sessionController.js:59]: That's right

File: server/src/models/quest.js
Comment (lancatlin) [reply server/src/models/quest.js:1]: [String] is array of string

**Review by Annie04082020 — APPROVED** [review]

**Pull request merged**